File: haskell/003/01.hs
-- The prime factors of 13195 are 5, 7, 13 and 29.
--
-- What is the largest prime factor of the number 600851475143 ?

main = putStrLn . show . maxFactor $ 600851475143

maxFactor :: Integer -> Integer
maxFactor n =
    head . filter (\d -> n `mod` d == 0) . reverse . takeWhile (<= max) $ primes
    where
      max = floor . sqrt . fromInteger $ n


primes = 2:3:go 5 [] (tail primes)
    where
      divisibleBy d n = n `mod` d == 0
      go start ds (p:ps) = let pSq = p*p in
          foldr (\d -> filter (not . divisibleBy d)) [start, start + 2..pSq - 2] ds
          ++ go (pSq + 2) (p:ds) ps

File: scala/003/01.scala
import scala.Math._

// http://mathworld.wolfram.com/DirectSearchFactorization.html
//
// What is the largest prime factor of the number 600851475143 ?

// produces a list of divisors of x, in decreasing order, excluding x
def divisors(x: Long): Seq[Int] = {
    // start at sqrt of x, since that's the largest possible divisor
    val r = round(floor(sqrt(x))).toInt to 1 by -1
    // identify divisors
    r.filter(x % _ == 0)
}

// a number is prime iff it's only divisors are itself and 1
def isPrime(x: Long): Boolean = {
    divisors(x).length == 1
}

// pull the first divisor that's prime (also the largest, since
// divisors returns in decreasing order
val biggun = 600851475143L
println(divisors(biggun).filter(isPrime(_)).first)
